### PR TITLE
Remove `g_glip` and `class GrpcLibraryInterface`

### PR DESCRIPTION
**Pull request by ralphchung**
This PR requires cherry picking before merging.

This PR aims to remove the pointer `g_glip` and `class GrpcLibraryInterface`.

Details:
* `grpc::GrpcLibraryCodegen` is moved to `grpc::GrpcLibrary` to replace the need for pointer `g_glip`.
* `grpc::GrpcLibraryInterface` is removed.

### CODE REVIEW

File: include/grpcpp/grpc_library.h [hunk 1]
```diff
@@ -16,13 +16,24 @@
...
-#endif  // GRPCXX_IMPL_CODEGEN_GRPC_LIBRARY_H
+/// Classes that require gRPC to be initialized should inherit from this class.
+class GrpcLibrary {
+ public:
+  explicit GrpcLibrary(bool call_grpc_init = true);
+  virtual ~GrpcLibrary();
+
+ private:
+  bool grpc_init_called_;
+  static internal::GrpcLibrary internal_grpc_library;
```
Comment (Vignesh2208) [inline include/grpcpp/grpc_library.h:34]: Why is there a static grpc::internal::GrpcLibrary ? It seems like its methods only call grpc_init() and grpc_shutdown(). Why not call them directly in grpc::GrpcLibrary constructor and destructor respectively ?

**Review by Vignesh2208 — CHANGES_REQUESTED** [review]
Looks mostly good to me.

File: include/grpcpp/grpc_library.h [hunk 2]
```diff
@@ -1,6 +1,6 @@
 /*
  *
- * Copyright 2018 gRPC authors.
+ * Copyright 2022 gRPC authors.
```
Comment (Vignesh2208) [inline include/grpcpp/grpc_library.h:3]: Please switch to newer style C++ header/copyright comments in this the file (e.g, src/core/lib/event_engine/iomgr_engine/ev_epoll1_linux.cc)

**Comment by ctiller** [comment]
Just noting here: this will need an internal cherry-pick, as there are internal usages of this header.

File: include/grpcpp/grpc_library.h (hunk 2 above)
Comment (ralphchung) [reply include/grpcpp/grpc_library.h]: Done

**Review by ctiller — APPROVED** [review]

File: BUILD [hunk 3]
```diff
@@ -6700,6 +6699,7 @@ grpc_cc_library(
     tags = ["nofixdeps"],
     visibility = ["@grpc:public"],
     deps = [
+        "grpc++_base",
```
Comment (ralphchung) [inline BUILD:6702]: I think this is a better place to bring the dependency given that we are removing `grpc++_codegen_base` and `grpc++_codegen_base_src` in the future.

Also, this change makes the blaze rule in Google internal not need to add the dependency.

**Review by Vignesh2208 — APPROVED** [review]

**Review by Vignesh2208 — CHANGES_REQUESTED** [review]
Actually there seem to be some bazel ASAN failures. Can you verify ?

File: include/grpcpp/grpc_library.h [hunk 4]
```diff
@@ -0,0 +1,43 @@
+// Copyright 2022 The gRPC Authors
```
Comment (ctiller) [inline include/grpcpp/grpc_library.h:1]: Lets move this to impl/... I don't think it should be public api

File: include/grpcpp/impl/grpc_library.h [hunk 5]
```diff
@@ -24,25 +24,34 @@
...
-class GrpcLibrary final : public GrpcLibraryInterface {
+
+/// Classes that require gRPC to be initialized should inherit from this class.
+class GrpcLibrary {
  public:
-  void init() override { grpc_init(); }
-  void shutdown() override { grpc_shutdown(); }
+  explicit GrpcLibrary(bool call_grpc_init = true) : grpc_init_called_(false) {
+    if (call_grpc_init) {
+      grpc_init();
```
Comment (ralphchung) [inline include/grpcpp/impl/grpc_library.h:37]: It seems calling `grpc_core::initInternally()` here might experience race condition. @ctiller do you have any suggestion on this?

Comment (ralphchung) [reply include/grpcpp/impl/grpc_library.h:37]: @ctiller suggested use `grpc_init()` instead.

**Review by Vignesh2208 — APPROVED** [review]

**Pull request merged**